### PR TITLE
Allow setting of custom websocket headers

### PR DESCRIPTION
**Pull request by agentgonzo**
Add the -websocketHeader flag to add additional custom headers to the connection request.

These additional headers are then used when connecting to jenkins via websockets.

Use-case: You have a reverse proxy in front of jenkins that requires custom headers to authenticate/route correctly.

Supercedes #455 

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/hudson/remoting/jnlp/Main.java
Comment (jglick) [inline src/main/java/hudson/remoting/jnlp/Main.java:311]: I do not suppose we need to imitate KK’s unusual formatting in new code, but does not matter!

**Review by jglick — APPROVED** [review]
Can I assume this has been tested in context somehow?

File: src/main/java/hudson/remoting/jnlp/Main.java
Comment (agentgonzo) [reply src/main/java/hudson/remoting/jnlp/Main.java:311]: I tend to favour consistency over correctness. Given there are a bunch of similar lines above/below it, I'd preder to keep that block consistent (either mimicking KK's style or replacing it all)

File: src/main/java/hudson/remoting/Engine.java
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Engine.java:354]: Javadoc would be nice, because this endpoint might be used on both server and client sides by API users

File: src/main/java/hudson/remoting/jnlp/Main.java
Comment (oleg-nenashev) [reply src/main/java/hudson/remoting/jnlp/Main.java:311]: We can refactor it in a follow-up if anyone has time

**Review by oleg-nenashev — APPROVED** [review]
Thanks @agentgonzo ! Looks good. 


We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

File: src/main/java/hudson/remoting/Engine.java
Comment (agentgonzo) [reply src/main/java/hudson/remoting/Engine.java:359]: @oleg-nenashev added.

**Comment by jeffret-b** [comment]
Try another build.

File: src/main/java/hudson/remoting/jnlp/Main.java
Comment (jeffret-b) [reply src/main/java/hudson/remoting/jnlp/Main.java:311]: We've been trying to move Remoting code to follow a more standard formatting style. But, this particular area hasn't been updated yet, so I suppose it's okay to follow the local formatting.

**Comment by jeffret-b** [comment]
We need a successful build, or a viable explanation for why we can ignore the failures.

**Comment by jvz** [comment]
That looks like an infra issue of some sort. It's a maven error.

**Comment by jeffret-b** [comment]
Probably. I'm still not comfortable with merging if we don't have a
successful build.

**Comment by jvz** [comment]
Right. Let's wait for the infra to repair itself first.

**Comment by abayer** [comment]
I've tested this manually and it works - the headers get added to the websocket connection properly.

**Comment by abayer** [comment]
(I also reran the build and that worked, so woo)

**Comment by agentgonzo** [comment]
Great, thanks @abayer. Definitely worth getting a green build before merging.

**Comment by agentgonzo** [comment]
I can also confirm from manual testing that the headers get added to the websocket connection and it works as expected.

**Review by jeffret-bg — APPROVED** [review]

**Pull request merged**
